driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -210,11 +301,17 @@ public BooleanExpression isArray() {
         return new MqlExpression<>(astWrapped("$isArray"));
     }
 
-    @SuppressWarnings("unchecked") // TODO
+    /**
+     * checks if array (but cannot check type)
+     * user asserts array is of type R
+     *
+     * @param or
+     * @return
+     * @param <R>
+     */
     @Override
-    public ArrayExpression<Expression> isArrayOr(final ArrayExpression<? extends Expression> or) {
-        // TODO it seems that ArrEx<T> does not make sense here
-        return (ArrayExpression<Expression>) this.isArray().cond(this.assertImplementsAllExpressions(), or);
+    public <R extends Expression> ArrayExpression<R> isArrayOr(final ArrayExpression<R> or) {
```

Comment:
Note the signature change. Previously, this was safer (by returning Array<Ex>), since it did not rely on the user asserting the type. But when I needed to use it in `getArray`, it was not useful enough. This might have been resolved with a cast, but I assumed that users would run into the same issue.

Reply:
We discussed this. It's sad that there is no way to make the API fully type-safe without significantly affecting the performance. Here it's because we are making the type of array elements part of the array type, which is the case neither in the BSON type system and in the MQL type system.